pretest.php

```diff
@@ -0,0 +1,12 @@
+<?php
+/**
+ * This script runs before the test command is executed.
+ *
+ * @package performance
+ */
+
+// On local environment.
+if ( ! getenv( 'WORDPRESS_DB_USER' ) ) {
```

Comment:
Or what about checking for all of the required environment variables, and if they aren't all set, then the error message could mention which one(s) are missing in case someone is wanting to run the tests from their machine without Docker. For example, the error message could be like:

> Please use 'npm run test-php' instead of running the composer test command directly.
>
> If you are intending to run PHPUnit directly, the following environment variables are missing: WORDPRESS_DB_USER, WORDPRESS_DB_PASSWORD.

Or alternatively, maybe it would make more sense to put this error message in the [`bootstrap.php`](https://github.com/WordPress/performance/blob/trunk/tests/bootstrap.php) instead of introducing this `pretest.php` file?

For example, before this line:

https://github.com/WordPress/performance/blob/2b2bebcb9a6285fac8653d078c1d49588f3c6b70/tests/bootstrap.php#L35

It could do:

```php
if ( ! file_exists( $_test_root . '/includes/functions.php' ) ) {
    echo "It seems your environment is not set up for running tests. (Unable to determine the test root.) Please use 'npm run test-php' instead.\n";
    exit( 1 );
}
```

Reply:
Hey, thanks for the input.

The first method seems easy enough, I can check for all required environment variables and show the missing error accordingly in `pretest.php`.

But I think the second approach would be better since it would integrate in the existing code without creating a new file/command.

I noticed that `file_exists( $_test_root . '/includes/functions.php' )` is true even on local, so I was thinking of adding all required environment variables check in the `bootstrap.php`.

Something like this:
```php
$required_constants = array(
    // required env variables.
);

$missing_constants = array();

foreach ( $required_constants as $constant ) {
    // Check if the environment variable is not set
    // if not add it to missing_constants array
}

if ( $missing_constants ) {
    // If there are missing constants print error
    exit( 1 );
}
```

The the error would be like this:
```bash
> phpunit '--verbose' '--testsuite' 'auto-sizes'
Please use 'npm run test-php' instead of running the composer test command directly.
If you are intending to run PHPUnit directly, the following environment variables are missing: WORDPRESS_DB_PASSWORD, WORDPRESS_DB_HOST, WORDPRESS_DB_USER, WORDPRESS_DB_NAME, PHP_VERSION, WP_TESTS_DIR
Script phpunit handling the test event returned with error code 1
Script @test --verbose --testsuite auto-sizes was called via test:auto-sizes
Script @test:auto-sizes was called via test:plugins
```

Reply:
It looks like on local I have the `WP_PHPUNIT__DIR` environment variable being set by Composer. This is due to the `wp-phpunit/wp-phpunit` package which has a `vendor/wp-phpunit/wp-phpunit/__loaded.php` file that is doing:

```php
putenv( sprintf( 'WP_PHPUNIT__DIR=%s', dirname( __FILE__ ) ) );
```

So yeah, that `file_exists()` check won't work _if_ a user had first done `composer install`. Still, I think it should include both that `file_exists()` check in addition to the check for the required environment variables.

Reply:
Agreed, added both checks in the [latest commit](https://github.com/WordPress/performance/blob/48a17f5bddfd785d02e746795d0b80922cf9df45/tests/bootstrap.php#L35-L66).